File: typescript/agentkit/src/action-providers/erc20/schemas.ts
import { z } from "zod";

/**
 * Input schema for transfer action.
 */
export const TransferSchema = z
  .object({
    amount: z.custom<bigint>().describe("The amount of the asset to transfer"),
    contractAddress: z.string().describe("The contract address of the token to transfer"),
    destination: z.string().describe("The destination to transfer the funds"),
  })
  .strip()
  .describe("Instructions for transferring assets");

/**
 * Input schema for get balance action.
 */
export const GetBalanceSchema = z
  .object({
    contractAddress: z
      .string()
      .describe("The contract address of the token to get the balance for"),
  })
  .strip()
  .describe("Instructions for getting wallet balance");

File: python/coinbase-agentkit/coinbase_agentkit/action_providers/erc20/schemas.py
"""Schemas for the ERC20 action provider."""

from pydantic import BaseModel, Field, field_validator

from .validators import wei_amount_validator


class GetBalanceSchema(BaseModel):
    """Schema for getting the balance of an ERC20 token."""

    contract_address: str = Field(
        ...,
        description="The contract address of the token to get the balance for",
    )


class TransferSchema(BaseModel):
    """Schema for transferring ERC20 tokens."""

    amount: str = Field(description="The amount of the asset to transfer in wei")
    contract_address: str = Field(description="The contract address of the token to transfer")
    destination: str = Field(description="The destination to transfer the funds")

    @field_validator("amount")
    @classmethod
    def validate_wei_amount(cls, v: str) -> str:
        """Validate wei amount."""
        return wei_amount_validator(v)
